Claude code guide (#1337)
checkly/docs.checklyhq.com: site/content/guides/claude-code-monitoring.md

---
title: Add in Depth Monitoring With Checkly and Claude Code
description: >-
  By integrating AI into your development workflow, you can generate valid Checkly constructs—including URL monitors and Playwright-based browser checks—with minimal effort.
author: Nočnica Mellifera
avatar: 'images/avatars/nica-mellifera.png'
tags:

>   - FAQ

Reviewer: Can we just go with `AI`? For whatever reason, this renders super strange in the guide section.

<img width="792" height="306" alt="grafik" src="https://github.com/user-attachments/assets/49522d2b-27de-4729-860b-7fae6dff31a8" />

Author: FAQ AI 😆 fixed

  - AI
---
AI-assisted coding promises to massively upgrade developer productivity, and with Checkly’s model of [Monitoring as Code](https://www.checklyhq.com/blog/monitoring-as-code-in-your-sdlc/), you can create monitoring coverage for all your services in minutes instead of days. This guide will show you how to create an AI development environment that lets you create and deploy all types of Checkly monitoring.

## Setup

If you haven’t already, start by [installing the Checkly CLI](https://www.checklyhq.com/docs/cli/installation/), then create a repository that will be the ‘project’ that contains all your Checkly monitors managed as code. If you don’t already have a project, create one with:

```bash
npm create checkly@latest
```

Next, install [Claude Code](https://www.anthropic.com/claude-code) globally with 

```bash
npm install -g @anthropic-ai/claude-code
```

and from the root folder of your project and start Code with the `claude` command.

You’ll be prompted to sign in to or sign up for an Anthropic account.

> I recommend using [the Visual Studio Code plugin for Claude](https://marketplace.visualstudio.com/items?itemName=anthropic.claude-code), which will make it easier to connect Claude Code with the VS Code file browser with Claude’s context (the VS Code plugin does a lot more than just context, this is just the most convenient feature we’ll use for this tutorial).

Reviewer: ```suggestion
I recommend using [the Visual Studio Code plugin for Claude](https://marketplace.visualstudio.com/items?itemName=anthropic.claude-code), which will make it easier to provide Claude Code with the VS Code file browser context (the VS Code plugin does a lot more than just context, this is just the most convenient feature we’ll use for this tutorial).
```

This sentence reads a little funky.

### Add the Playwright MCP

You can give Claude Code access to a browser by adding the [Playwright MCP](https://github.com/microsoft/playwright-mcp) to your local configuration with 

```bash
claude mcp add playwright npx @playwright/mcp@latest 
```
This will let us ask Claude Code to do things like crawling a particular website and acting on what it finds by opening the site in a browser.

> ## Give Claude Code Some Context

Reviewer: Is the casing of the headings inconsistent? E.g., we have:

- Give Claude Code Some Context
- `Create new URL Monitors` (small `new`)
- `Create Playwright Synthetics checks with Claude Code` (small `checks` and `with`)

I'm always confused by all the different title cases. :D Is this how we want to do things

Author: Ahh the great American/German divide, the European language that cares the most vs. the one that cares the least about capitalization. The actual standard should be "less than 4 letters and not a proper name: not capitalized" so 'Checks' should be. Will correct.

![Feedback on the terminal](/guides/images/claude-monitoring-01.png)

## Create Playwright Synthetics checks with Claude Code

It would be nice to have Claude Code automatically create the Playwright scripts we need to test our site's features. However, without careful controls, any 2025-era coding agent tends to write Playwright code that is either out of date or doesn’t follow best practices. The best way to get high-quality output is through some prompt engineering and careful provision of context.

> ### 1. Create at least one test with Playwright Codegen

Reviewer: I think I mentioned this before, but it doesn't feel right to me. 

Why would I reach for codegen in the first place? I could instruct Claude to read the docs. And the article also mentions the PWT MCP server in the opening. I doubt that 7 lines of PWT code make a big difference in context engineering here. 🫣

Author: Can revise the heading. In my experiments the currently open file is a critical part of the prompt, as it's fed in verbatim rather than summarized.

Author: We may be talking across each other, I'll make myself available for a huddle to talk through it. Here's what I tested:

* Just scanning my project with /init
* Running /init, adding the Checkly Rules to CLAUDE.md
* Running /init, adding the Checkly Rules, and adding a single check as context

In the final bullet, I got working tests 4 out of 5 times, in the other runs it was worse, more like 1 out of 5 runs were working checks, I had to tweak my request like 'give me working code, be sure to actually load the page in a browser' etc.

I haven't debugged this, but my understanding is that when you've got a code file open in Claude Code, its entire text (length permitting) is added to the prompt, rather than the summaries that appear in CLAUDE.md. *Sometimes* Claude Code will scan, find an existing check, and add it in full to the prompt, but not always.

This isn't scientific, and I'll be the first to admit that the final bullet point feels like a 'belt and braces' approach, but it is the process that I actually ended up using.

"Okay," you might say, "so you want to feed in a check file, why create one with Codegen and not use an existing one instead?"

There are some reasons I think this is a bit better technically, but more importantly **I'd like the reader to try Codegen if they haven't already**

-----
Sorry for the very long response, I was re-reading comments and worried that I seemed obtuse or stubborn on this point.

Reviewer: > In the final bullet, I got working tests 4 out of 5 times, in the other runs it was worse, more like 1 out of 5 runs were working checks, I had to tweak my request like 'give me working code, be sure to actually load the page in a browser' etc.

Are you really sure the reason was that you had 15 more lines in your context? It seems very unlikely that these few lines make a huge difference. Maybe your prompts were better?

And sorry for being a bit more stubborn here, however, as pointed out, I've seen it working without providing any PWT code. "Record an example to let AI replicate it" just doesn't feel great to me in this case. 

> There are some reasons I think this is a bit better technically, but more importantly I'd like the reader to try Codegen if they haven't already

I don't think this article is the place to more than hint towards codegen. This article is about AI. Or compare the two. But pointing out that you recommend `codegen` to get the AI approach working feels very backwards to me.

----

Ja, so generally, I don't think using Codegen here is helping sell the AI topic very well. 

- As I read it, I immediately question why I should use AI in the first place.
- Using `codegen` to add a few lines of PWT code to the context feels strange. The LLM should point to the docs and/or adjust what comes with the `checkly.rules`.

We can capture our behavior in the browser to script a test with [Playwright Codegen](https://www.checklyhq.com/learn/playwright/codegen/), available either as a standalone utility, browser plugin, or VS Code plugin. Once [Codegen is set up](https://www.checklyhq.com/learn/playwright/codegen/), start recording and record the following:

- Navigate to your homepage.
- Browse through multiple pages.

> - Assert the visibility of 1-2 page elements.

Author: Yeah that *is* a cromulent way to do this, see above comment for my reasoning.

```ts
  await expect(page.getByText('Fantasy').first()).toBeVisible();
```

If you’re having trouble reading the generated tests, take a look at [Checkly’s Playwright documentation](https://www.checklyhq.com/learn/playwright/). Once Claude Code added a single `.first()` to my check, everything was passing:

![Feedback on the terminal](/guides/images/claude-monitoring-04.png)

## Deploy With the Checkly CLI

In our example, we want to create several checks at once, so it’s worth checking whether the configuration is properly formatted before deployment. Many problems would have been revealed while running `test` but if, for example, the [logical IDs](https://www.checklyhq.com/docs/cli/constructs-reference/#synthetic-checks) of current and existing checks are colliding, we’ll need to run `deploy` to detect that. Run the `deploy` command with the `-p` preview flag:

```bash
npx checkly deploy -p
```

This command won’t deploy anything, just parse our project and show us what will be created when we run the command without the preview flag.

> ![Feedback on the terminal](/guides/images/claude-monitoring-05.png)

Reviewer: Nit: this screenshot is slightly confusing. It shows 7 checks. This entire guide only discusses two.

```bash
npm create checkly@latest
```

Next, install [Claude Code](https://www.anthropic.com/claude-code) globally with 

```bash
npm install -g @anthropic-ai/claude-code
```

and from the root folder of your project and start Code with the `claude` command.

You’ll be prompted to sign in to or sign up for an Anthropic account.

I recommend using [the Visual Studio Code plugin for Claude](https://marketplace.visualstudio.com/items?itemName=anthropic.claude-code), which will make it easier to connect Claude Code with the VS Code file browser with Claude’s context (the VS Code plugin does a lot more than just context, this is just the most convenient feature we’ll use for this tutorial).

Optionally, from the Claude Code prompt, run `/init` to scan your project and create an initial `CLAUDE.md` context file.

### Add the Playwright MCP

You can give Claude Code access to a browser by adding the [Playwright MCP](https://github.com/microsoft/playwright-mcp) to your local configuration with 

```bash
claude mcp add playwright npx @playwright/mcp@latest 
```

> ## Give Claude Code Some Context

Reviewer 2: As someone who hasn’t created a rules file before, I felt a little lost at this step of the guide. It'll be helpful to get a few pointers on what we’re looking to do here. For example:

* I assume I can define config rules here like naming patterns or specifying which properties (like frequency, tags, assertions, etc.) should always be included for a check type?
* I guess you could also define rules when new monitor should be created, like whenever a new API endpoint is added, Claude can suggest creating an API check for it?

If this feels redundant from your POV please ignore 🙂

Reviewer: Agreed. :)

Author: added a whole explanation and tweaked the order to feel a bit more logical.

```bash
mkdir -p .claude &&
curl -o .claude/checkly.rules.md https://www.checklyhq.com/docs/ai/checkly.rules.md -L 
echo "- examine checkly.rules.md for code generation rules" >> .claude/CLAUDE.md
```

After this you’ll need to exit and restart Claude Code for it to pick up changes to our context file.

## Create new URL Monitors

We’ll start with something easy. A URL monitor that will check an HTTP endpoint and alert if it doesn’t receive the right status code. With the context set from the section above, try a command like:

> Create a new URL pinger for the site [https://danube-webshop.herokuapp.com/](https://danube-webshop.herokuapp.com/) that runs every five minutes and doesn’t follow redirects

After a few seconds, Claude Code will produce this:

```ts
import { Frequency, UrlMonitor, UrlAssertionBuilder } from 'checkly/constructs'

new UrlMonitor('danube-web-shop-pinger', {
  frequency: Frequency.EVERY_5M,
  name: 'Danube Web Shop Pinger',
  activated: true,
  request: {
    url: 'https://danube-web.shop/',
    skipSSL: false,

>     followRedirects: true,

Reviewer 2: ```suggestion
    followRedirects: false,
```

See prompt above "and doesn’t follow redirects"

  await expect(page.getByText('$9.95').first()).toBeVisible();
  await expect(page.getByRole('heading', { name: 'Top sellers' })).toBeVisible();
});
```

Claude Code will confirm before creating the check and spec files with the configuration and script, respectively.

You can prompt for multiple checks at the same time, as the output is relatively token-efficient. As you can imagine, this unlocks the creation of a large number of synthetic monitors relatively quickly. Another use case that may be useful for larger projects, try prompting Claude Code with something like:

> 
> 
> 
> update every `.check.ts` file for a check that touches [danube-web.shop](http://danube-web.shop/), and change
> the frequency to every 30 minutes.
> 

Not only did this request work, but I even got sensible feedback about the checks that currently had no frequency setting:

```bash
  The url.check.ts and api.check.ts files don't have individual frequency settings -
  they inherit the default frequency from the project configuration in
  checkly.config.ts.
```

Once we’re done creating new checks, it’s time to run them all with:

```bash
npx checkly test
```

> When I ran a number of generated tests, the only issue I ran into was a failure with the message: “Error: expect.toBeVisible: Error: strict mode violation: getByText('Crime & Thrillers') resolved to 2 elements:” As the error message implies, the issue is that this locator:

Reviewer: ```suggestion
When I ran a number of generated tests, the only issue I ran into was a failure with the message: `Error: expect.toBeVisible: Error: strict mode violation: getByText('Crime & Thrillers') resolved to 2 elements:` As the error message implies, the issue is that this locator:
```

```ts
  await expect(page.getByText('Fantasy')).toBeVisible();
```

Found multiple results on the page. There are a number of ways to fix this with Claude Code, we could prompt Claude Code with:

> revise locators in item-visibility.spec.ts to look at only the first result
> 

This would change every locator, which may or may not be what you want! You could also paste in the error message to Claude and ask it to fix the script. Whatever route you take the result will be an edit to:

```ts
  await expect(page.getByText('Fantasy').first()).toBeVisible();
```

If you’re having trouble reading the generated tests, take a look at [Checkly’s Playwright documentation](https://www.checklyhq.com/learn/playwright/). Once Claude Code added a single `.first()` to my check, everything was passing:

> ![Feedback on the terminal](/guides/images/claude-monitoring-04.png)

Author: replaced!

## Deploy With the Checkly CLI

> In our example, we want to create over 50 checks at once, so it’s worth checking whether the configuration is properly formatted before deployment. Many problems would have been revealed while running `test` but if, for example, the [logical IDs](https://www.checklyhq.com/docs/cli/constructs-reference/#synthetic-checks) of current and existing checks are colliding, we’ll need to run `deploy` to detect that. Run the `deploy` command with the `-p` preview flag:

Author: revised

While AI can handle much of the boilerplate work, human oversight remains essential for ensuring best practices, refining assertions, and maintaining high-quality monitoring scripts. By combining AI efficiency with developer expertise, teams can achieve comprehensive monitoring coverage in minutes rather than days.

## Further Reading

To learn more about Checkly’s capabilities, explore these resources:

- [Checkly CLI Documentation](https://www.checklyhq.com/docs/cli/) – Learn how to manage checks as code right from your command line.
- [Playwright Testing Guide](https://www.checklyhq.com/learn/playwright/) – Best practices for writing reliable browser checks.
- [What is monitoring as code?](https://www.checklyhq.com/guides/monitoring-as-code/) — A comprehensive guide to infrastructure-as-code principles for monitoring.
- [Checkly Community Slack](https://checklyhq.com/slack) – Join discussions and get support from the Checkly team and other users.

For more tutorials and advanced use cases, visit the [Checkly YouTube channel](https://www.youtube.com/@ChecklyHQ), where you can see great tutorials like this guide to the Playwright MCP server:

> {{< youtube id="kzQu0Y_Nxjk" title="Generating Playwright Tests With AI: Let's Try the New Playwright MCP Server!">}}

Author: fixed!

```

After a few seconds, Claude Code will produce this:

```ts
import { Frequency, UrlMonitor, UrlAssertionBuilder } from 'checkly/constructs'

new UrlMonitor('danube-web-shop-monitor', {
  frequency: Frequency.EVERY_5M,
  name: 'Danube Web Shop URL monitor',
  activated: true,
  request: {
    url: 'https://danube-web.shop/',
    skipSSL: false,
    followRedirects: false,
    assertions: [
      UrlAssertionBuilder.statusCode().equals(200),
    ]
  }
})
```

While we're writing this simplest of monitors, it's worth testing the limits of Claude Code's context for writing valid Checkly configuration. One thing we tested while writing this article was whether picking the wrong prompt would result in invalid Checkly construct code. The `Frequency` class here doesn't accept arbitrary values, so we wondered if the context provided in `CLAUDE.md` was enough for the tool to identify valid values for the check. We tried requesting a check that ran “every 17 seconds” and Claude Code prompted to run a `find` on the project to identify valid values for `frequency`. In the end, Claude Code did create valid code with this note in the process feed.

```{title="Reply from Claude Code:"}

>   Based on the search results, 17 seconds is NOT a valid frequency value for Checkly monitors. The closest valid option was selected: 20 seconds.

Reviewer: ```suggestion
Based on the search results, 17 seconds is NOT a valid frequency value for Checkly monitors. The closest valid option was selected: 20 seconds.
```

```

This would change every locator in the check, which may or may not be what you want! You could also paste in the error message to Claude and ask it to fix the script, or specify which locator you want to add `first()` to. Whatever route you take the result will be an edit to:

```ts
  await expect(page.getByText('Fantasy').first()).toBeVisible();
```
Once Claude Code added a single `.first()` to the checks written above, everything was passing. If you're having trouble reading the generated tests, take a look at [Checkly's Playwright documentation](https://www.checklyhq.com/learn/playwright/). 

### 4. Going Further: Edit Multiple Checks and Configurations

Uodating checks is another use case that may be useful for larger projects. Try prompting Claude Code with something like:

```{title="Prompt:"}
Update every `.check.ts` file for a check that touches [danube-web.shop](http://danube-web.shop/), and change the frequency to every 30 minutes.
```

Not only did this request work, but even produced sensible feedback about the checks that currently had no frequency setting:

```{title="Reply from Claude Code:"}

>   The url.check.ts and api.check.ts files don't have individual frequency settings - they inherit the default frequency from the project configuration in checkly.config.ts.

Reviewer: ```suggestion
The url.check.ts and api.check.ts files don't have individual frequency settings - they inherit the default frequency from the project configuration in checkly.config.ts.
```

```bash
npx checkly deploy -p
```

This command won't deploy anything, just parse our project and show us what will be changed when we run the command without the preview flag. When you first run `deploy -p` you should just see a long list of created checks in the deployment command. For this tutorial, here's the preview of a more complex project state:

```bash
Bundling project resources... ✅

Create:
    BrowserCheck: __checks__/crime-category.spec.ts
    BrowserCheck: __checks__/fantasy-category.spec.ts
    BrowserCheck: __checks__/homepage-elements.spec.ts
    BrowserCheck: __checks__/navigation-menu.spec.ts
    BrowserCheck: __checks__/search-functionality.spec.ts
    BrowserCheck: book-detail-visibility-check-v1
    BrowserCheck: signup-form-elements-check-v1

Delete:

    Check: danube-web-shop-monitor

Update and Unchanged:
    BrowserCheck: __checks__/homepage.spec.ts
```

Let's discuss the output of this preview command in detail, as it's critical to understand how Checkly handles deployments from your project:

> * Checks that have no .check.ts config file are listed by their filename and will be named for their file name in the Checkly web UI. In general, it's better to include a config file for every check since it gives you explicit control of the check's logical ID. We've left out the config for five checks for demonstration purposes.

Reviewer: ```suggestion
* Checks that have no `.check.ts` config file are listed by their filename and will be named for their file name in the Checkly web UI. In general, it's better to include a config file for every check since it gives you explicit control of the check's logical ID. We've left out the config for five checks for demonstration purposes.
```

Can we be consistent and have all file refs in inline code? 🫣

* Checks with a config file are shown by their logical ID. 
* Checks previously deployed from this project that are not present in this deployment will be deleted.

> * Checks with a matching logical ID from a previous deployment will be listed as "Update and Unchanged"

Reviewer: ```suggestion
* Checks with a matching logical ID from a previous deployment will be listed as "Update and Unchanged".
```

## Conclusion

AI-assisted coding tools like Claude Code can significantly accelerate the creation and deployment of monitoring checks with Checkly's Monitoring as Code approach. By integrating AI into your development workflow, you can generate valid Checkly constructs—including URL monitors and Playwright-based browser checks—with minimal effort. One of the superpowers that an AI workflow can unlock is the *mass updating of many files in your project.* Updating a large number of test files can go much faster with the help of Claude Code. 

While AI can handle much of the boilerplate work, human oversight remains essential for ensuring best practices, refining assertions, and maintaining high-quality monitoring scripts. By combining AI efficiency with developer expertise, teams can achieve comprehensive monitoring coverage in minutes rather than days.

### Further Reading

Throughout this tutorial, we added a few lines to the CLAUDE.md context file. In all, those editions were (rewritten for clarity):

```markdown
## Memories and Best Practices
- Follow the rules in .claude/checkly.rules.md
### When writing checkly playwright tests

> -  load the target page in a browser

Reviewer: ```suggestion
- load the target page in a browser
```

### Add the Playwright MCP

You can give Claude Code access to a browser by adding the [Playwright MCP](https://github.com/microsoft/playwright-mcp) to your local configuration with 

```bash
claude mcp add playwright npx @playwright/mcp@latest 
```
This will let us ask Claude Code to do things like crawling a particular website and acting on what it finds by opening the site in a browser.

## Give Claude Code Some Context
Everything we write into the CLI interface is part of a 'prompt,' and by default, Claude Code will add a bit of information to what we submit as 'context' on that prompt. Ideally, our requests written into Claude Code would include context about our specific project: its structure, language, and even its purpose. Because of the `npm create` command we ran above, our folder already has a boilerplate version of a Checkly project. We can't feed in all of these files as context, but we can add a summary of our whole project as context by opening Claude Code and running the command:

```{title="Prompt:"}
/init
```

> Claude Code will scan the whole project, write a summary, and then store that summary in a local CLAUDE.md file. Claude Code will read that file and add it to our prompt every time it's run within this project from now on.

Reviewer: ```suggestion
Claude Code will scan the whole project, write a summary, and then store that summary in a local `CLAUDE.md` file. Claude Code will read that file and add it to our prompt every time it's run within this project from now on.
```

Let's add a bit more to this context, we want test files that make sense as a [Checkly construct](https://www.checklyhq.com/docs/cli/constructs-reference/). Let's add [Checkly's AI rules file](https://www.checklyhq.com/docs/ai//use-checkly-with-ai-ide/#claude-code) to our CLAUDE.md context file:

```bash
mkdir -p .claude &&
curl -o .claude/checkly.rules.md https://www.checklyhq.com/docs/ai/checkly.rules.md -L 
echo "- examine checkly.rules.md for code generation rules" >> .claude/CLAUDE.md
```

After this you'll need to exit and restart Claude Code for it to pick up changes to our context file.

## Create new URL Monitors

We'll start with something easy. A URL monitor that will check an HTTP endpoint and alert if it doesn't receive the right status code. With the context set from the section above, try a command like:

```{title="Prompt:"}
Create a new URL monitor for the site https://danube-webshop.herokuapp.com/ that runs every five minutes and doesn't follow redirects.
```

After a few seconds, Claude Code will produce this:

> ```ts

Reviewer: This is the only file code block with a file name header. Is this in purpose?

Author: Just trying to keep the visual clear, it's the only code block where we don't refer back to the title.

import { Frequency, UrlMonitor, UrlAssertionBuilder } from 'checkly/constructs'

new UrlMonitor('danube-web-shop-monitor', {
  frequency: Frequency.EVERY_5M,
  name: 'Danube Web Shop URL monitor',
  activated: true,
  request: {
    url: 'https://danube-web.shop/',
    skipSSL: false,
    followRedirects: false,
    assertions: [
      UrlAssertionBuilder.statusCode().equals(200),
    ]
  }
})
```

> While we're writing this simplest of monitors, it's worth testing the limits of Claude Code's context for writing valid Checkly configuration. One thing we tested while writing this article was whether picking the wrong prompt would result in invalid Checkly construct code. The Frequency class here doesn't accept arbitrary values, so we wondered what would happen since the [source file for Frequency](https://github.com/checkly/checkly-cli/blob/main/packages/cli/src/constructs/frequency.ts) won't be part of Claude's context. we tried requesting a check that ran “every 17 seconds” and Claude Code prompted to run a `find` on the project to identify valid values for `frequency`. In the end, Claude Code did create valid code with this note in the process feed.

Reviewer: ```suggestion
While we're writing this simplest of monitors, it's worth testing the limits of Claude Code's context for writing valid Checkly configuration. One thing we tested while writing this article was whether picking the wrong prompt would result in invalid Checkly construct code. The Frequency class here doesn't accept arbitrary values, so we wondered what would happen since the [source file for Frequency](https://github.com/checkly/checkly-cli/blob/main/packages/cli/src/constructs/frequency.ts) won't be part of Claude's context. We tried requesting a check that ran “every 17 seconds” and Claude Code prompted to run a `find` on the project to identify valid values for `frequency`. In the end, Claude Code did create valid code with this note in the process feed.
```

> While we're writing this simplest of monitors, it's worth testing the limits of Claude Code's context for writing valid Checkly configuration. One thing we tested while writing this article was whether picking the wrong prompt would result in invalid Checkly construct code. The Frequency class here doesn't accept arbitrary values, so we wondered what would happen since the [source file for Frequency](https://github.com/checkly/checkly-cli/blob/main/packages/cli/src/constructs/frequency.ts) won't be part of Claude's context. we tried requesting a check that ran “every 17 seconds” and Claude Code prompted to run a `find` on the project to identify valid values for `frequency`. In the end, Claude Code did create valid code with this note in the process feed.

Reviewer: Nit: `Frequency` is included 18 times in the Checkly AI context that was added, though. Maybe that's worth pointing out. I could think of something like

> while `Frequency` is used in the provided Checkly AI guidelines which live now in the `CLAUDE.md`, it's never read the `Frequency` class directly...

or something.

```{title="Reply from Claude Code:"}
  Based on the search results, 17 seconds is NOT a 
  valid frequency value for Checkly monitors. The
  closest valid option was selected: 20 seconds.
```

## Create Playwright Synthetics Checks With Claude Code

> With our context set up, and a browser available via the Playwright MCP, we can create more complex playwright scripts for browser monitoring of our site. While we should expect a coding agent to produce simpler uptime monitors with very high consistency, Playwright tests are a much more flexible testing framework, and as such a 'good' test will depend on your exact user needs. With specific instructions and clear prompting, you can get consistent performance of Claude Code writing our Browser checks.

Reviewer: > While we should expect a coding agent to produce simpler uptime monitors with very high consistency, Playwright tests are a much more flexible testing framework, and as such a 'good' test will depend on your exact user needs. 

This reads a little bumpy... could we rephrase?

### 1. Start With a single test

First you'll want to write a single check that does straightforward browsing and checking for page elements. Once we know this check is working, we can extend this into broader scales of monitoring.

> ```{title="Prompt:"}

Reviewer: This won't generate a test, though, will it? Suggestions, why don't we link to the blog post.

https://www.checklyhq.com/blog/generate-end-to-end-tests-with-ai-and-playwright/

And change the wording to something like 

> With some context engineering, you can instruct Claude to navigate a page while recording the taken actions to generate a Playwright test. Read more about this approach on the blog.

And then start the prompt block with

```
You are a Playwright test generator, ...

- Navigate
- Browser
- Assert
```
```

Author: Here's the shortened prompt I went with, it performed pretty well (given the other steps above), but you might want to add more of the ideal prompt from the blog, feel free to commit an edit:

```{title="Prompt:"}
You are a Playwright test generator and an expert in TypeScript, Frontend development, and Playwright end-to-end testing.
- You are given a scenario, and you need to generate a Checkly spec.ts test for it.
- Navigate to the homepage https://danube-web.shop/
- Browse through multiple pages.
- Assert the visibility of 1-2 page elements.
```

    
    await page.getByText('Parry Hotter').click()    
    await expect(page.getByRole('heading', { name: 'Parry Hotter' })).toBeVisible()
    await expect(page.getByRole('button', { name: 'Add to cart' })).toBeVisible()
    await expect(page.locator('img').first()).toBeVisible()
  })
})
```

Lets make sure our script is valid by running this test:

```bash
npx checkly test book-detail
```
*Without an exact file path, the `checkly test` command will run any test that matches this pattern.*

and making sure the test passes.

```bash
Bundling project resources... ✅

Running 1 checks in eu-west-1.

__checks__/book-detail-visibility.spec.ts
  ✔ book-detail-visibility.spec.ts (7s)

1 passed, 1 total
```

*Like always with the `checkly test` command, this test isn't running from our workstation but rather from the Checkly's monitoring infrastructure.*

Next we'll want to create a `.check.ts` file for this browser check, which sets some check-specific settings and assigns the check's logical ID. 

```{title="Prompt:"}
write a .check.ts file for book-detail-visibility that runs the check from a US and EU region and defines a linear retry strategy 
```

> This should result in a complete `.spec.ts` file generated by Claude, implementing the Checkly construct configuration. For better understanding of the configuration, you can ask for code comments:

Reviewer: ```suggestion
This should result in a complete `.check.ts` file generated by Claude, implementing the Checkly construct configuration. For better understanding of the configuration, you can ask for code comments:
```

    enabled: false,
    // Percentage threshold for escalation (when enabled)
    percentage: 10,
  }),
  // Retry strategy configuration for failed checks
  retryStrategy: RetryStrategyBuilder.linearStrategy({
    // Initial wait time before first retry in seconds
    baseBackoffSeconds: 30,
    // Maximum number of retry attempts
    maxRetries: 2,
    // Maximum total time for all retries in seconds
    maxDurationSeconds: 300,
    // Whether retries should use the same region as original check
    sameRegion: true,
  }),
  // Whether to run checks in multiple locations simultaneously
  runParallel: true,
})
```

These settings are all optional, and any not specified will default back to the account or project default, but it's good to include them, especially the logical ID, which will make it easy to update this check with later deployments.

Later in this tutorial we'll check the consistency of this config file with the `deploy` command. For now you can move on to creating more checks.
### 2. Add our new Check to Claude's Context

With this check created, add their references to Claude's context by giving Claude Code the prompt:

```{title="Prompt:"}

> update my CLAUDE.md file with the new code in /__checks__

Reviewer: The `CLAUDE.md` already includes a ton of checks thanks to the checkly AI file. Do we need to mention this?

```

Let's take one additional step, add the following lines to your updated CLAUDE.md file:

```markdown
- When writing Playwright, don't set locators equal to const, rather just perform expect tests directly on locators
- When writing a spec.ts file, don't use locators based on CSS class
```

Pro tip: from the Claude Code prompt, just hit octothorp "#" to add to Claude's memory.

### 3. Use Claude Code to create multiple Browser Checks

Now that we've got a working test and, let's let Claude Code create another check for us:

```{title="Prompt:"}
Create a new browser check for https://danube-web.shop/ that clicks on a single button then checks the visibility of 10 different elements on the page
```

> In order to scan the URL provided, Claude Code will confirm that it can use the Playwright MCP to open a browser and access the site. The result is a nice clean test written to the spec:

Reviewer: Will it do that even though Playwright MCP should have been used with the earlier example?

> ```ts {title="signup-form-elements.spec.ts"}

Reviewer: How is this script matching the prompt you mentioned above?

> Create a new browser check for https://danube-web.shop/ that clicks on a single button then checks the visibility of 10 different elements on the page

Author: Clarified why a general prompt gets us a pretty specific test. If this still feels too 'out of left field' then we can revise it to click on another button.

  test('click signup button and verify ten form elements are visible', async ({ page }) => {
    await page.goto('https://danube-web.shop/')
    
    await page.getByRole('button', { name: 'Sign up' }).click()
    
    await expect(page.getByRole('dialog')).toBeVisible()
    await expect(page.getByRole('textbox', { name: 'Name' }).first()).toBeVisible()
    await expect(page.getByRole('textbox', { name: 'Surname' })).toBeVisible()
    await expect(page.getByRole('textbox', { name: 'Email' })).toBeVisible()
    await expect(page.getByRole('textbox', { name: 'Password' })).toBeVisible()
    await expect(page.getByRole('textbox', { name: 'Company (optional)' })).toBeVisible()
    await expect(page.getByRole('radio', { name: 'Myself' })).toBeVisible()
    await expect(page.getByRole('radio', { name: 'My business' })).toBeVisible()
    await expect(page.getByRole('checkbox', { name: 'I would like to receive promotional emails' })).toBeVisible()
    await expect(page.getByRole('checkbox', { name: 'I have read and accept the privacy policy' })).toBeVisible()
    await expect(page.getByRole('button', { name: 'Register' })).toBeVisible()
  })
})

> });

Reviewer: This part `});` is too much or? 🤔

```ts
  await expect(page.getByText('Fantasy')).toBeVisible();
```

Found multiple results on the page. There are a number of ways to fix this with Claude Code. We could prompt Claude Code with:

```{title="Prompt:"}
revise locators in item-visibility.spec.ts to look at only the first result
```

This would change every locator, which may or may not be what you want! You could also paste in the error message to Claude and ask it to fix the script. Whatever route you take the result will be an edit to:

```ts
  await expect(page.getByText('Fantasy').first()).toBeVisible();
```
If you're having trouble reading the generated tests, take a look at [Checkly's Playwright documentation](https://www.checklyhq.com/learn/playwright/). Once Claude Code added a single `.first()` to the checks written above, everything was passing. This issue proved frequent enough that we added the following to Claude Code's permanent memory:

```{title="Prompt:"}

> # When writing a Playwright check, if you find multiple elements that match a locator, make sure the locator includes .first() to take the first result.

Reviewer: This is a very **bold** statement to have in context. I'm not sure if i'd feel comfortable recommending this, because it'll drive you nuts if you don't want to select the first element. :D 🫣

```
This solved the general issue for new browser checks created by Claude Code.

### 4. Going Further: Edit Multiple Checks and Configurations

Another use case that may be useful for larger projects, try prompting Claude Code with something like:

```{title="Prompt:"}
update every `.check.ts` file for a check that touches [danube-web.shop](http://danube-web.shop/), and change the frequency to every 30 minutes.
```

Not only did this request work, but even produced sensible feedback about the checks that currently had no frequency setting:

```{title="Reply from Claude Code:"}
  The url.check.ts and api.check.ts files don't have individual frequency settings -
  they inherit the default frequency from the project configuration in
  checkly.config.ts.
```

Once we're done creating new checks, it's time to run them all.

## Deploy With the Checkly CLI

> In our example, we want to deploy a couple of checks at once, so it's worth checking whether the configuration is properly formatted before deployment. Many problems would have been revealed while running `test`, but if, for example, the [logical IDs](https://www.checklyhq.com/docs/cli/constructs-reference/#synthetic-checks) of current and existing checks are colliding, we'll need to run `deploy` to detect that. Run the `deploy` command with the `-p` preview flag:

Reviewer: The statement that only `deploy` detects colliding logical ids is incorrect.

<img width="703" height="138" alt="image" src="https://github.com/user-attachments/assets/509a75c1-dba5-406e-88f8-f33f8c623659" />

Author: whoa.gif

Didn't realize it would detect if there was collision not in your current tests but also with previously deployed ones. Edited to clarify we're really checking for project consistency.

```bash
npx checkly deploy -p
```

This command won't deploy anything, just parse our project and show us what will be created when we run the command without the preview flag. When you first run `deploy -p` you should just see a long list of created tests in the deployment command. For this tutorial, here's the preview of a more complex project state:

```bash
Bundling project resources... ✅

Create:
    BrowserCheck: __checks__/crime-category.spec.ts
    BrowserCheck: __checks__/fantasy-category.spec.ts
    BrowserCheck: __checks__/homepage-elements.spec.ts
    BrowserCheck: __checks__/navigation-menu.spec.ts
    BrowserCheck: __checks__/search-functionality.spec.ts
    BrowserCheck: book-detail-visibility-check-v1
    BrowserCheck: signup-form-elements-check-v1

Delete:

    Check: danube-web-shop-monitor

Update and Unchanged:
    BrowserCheck: __checks__/homepage.spec.ts
```

Let's discuss the output of this preview command in detail, as it's critical to understand how Checkly handles deployments from your project:

> * Checks that have no .check.ts config file are listed by their filename and will be named for their file name in the Checkly web UI. In general, it's better to include a config file for every check since it leaves you in explicit control of the check's logical ID. We've left out the config for five checks for demonstration purposes.

Reviewer: ```suggestion
* Checks that have no `.check.ts` config file are listed by their filename and will be named for their file name in the Checkly web UI. In general, it's better to include a config file for every check since it leaves you in explicit control of the check's logical ID. We've left out the config for five checks for demonstration purposes.
```

While AI can handle much of the boilerplate work, human oversight remains essential for ensuring best practices, refining assertions, and maintaining high-quality monitoring scripts. By combining AI efficiency with developer expertise, teams can achieve comprehensive monitoring coverage in minutes rather than days.

### Further Reading

Throughout this tutorial, we added a few pointers to the CLAUDE.md context file. In all, those editions were (rewritten for clarity):

```markdown
## Memories and Best Practices
- Follow the rules in .claude/checkly.rules.md
### When writing checkly playwright tests
-  load the target page in a browser
- don't set locators equal to const, rather just perform expect tests directly on locators
- don't use locators based on CSS class
- if you find multiple elements that match a locator, make sure the locator includes 

>   .first() to take the first result.

Author: removed this line per above
